riscv_isa.cpp

```diff
@@ -1,489 +1,810 @@
+  dbg_printf("ADDI r%d, r%d, %d\n", rd, rs1, imm);
   if ((rd == 0) && (rs1 == 0) && (imm == 0)) {
-    ac_pc = ac_pc + 4;
     dbg_printf("NOP executed!");
-  } else {
+  } 
+  else {
     int sign_ext;
-    if ((imm >> 11) == 1)
+    if (imm4 == 1)
       sign_ext = imm | 0xFFFFF000;
     else
       sign_ext = imm;
     RB[rd] = RB[rs1] + sign_ext;
-    dbg_printf("Result = %#x\n", RB[rd]);
+    dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+    dbg_printf("imm = %d\n", sign_ext);
+    dbg_printf("Result = %d\n\n", RB[rd]);
   }
 }
 
+
+// Instruction SLTI behavior method.
 void ac_behavior(SLTI) {
-  dbg_printf("SLTI r%d, r%d, %d", rd, rs1, imm);
+  int imm;
+  imm = (imm4 << 11) | (imm3 << 5) | (imm2 << 1) | imm1;
+  dbg_printf("SLTI r%d, r%d, %d\n", rd, rs1, imm);
   int sign_ext;
-  if ((imm >> 11) == 1)
+  if (imm4 == 1)
     sign_ext = imm | 0xFFFFF000;
   else
     sign_ext = imm;
   if (RB[rs1] < sign_ext)
     RB[rd] = 1;
   else
     RB[rd] = 0;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("RB[rs1] = %d\n", RB[rs1]);
+  dbg_printf("imm = %d\n", sign_ext);
+  dbg_printf("Result = %d\n\n", RB[rd]);
 }
 
+
+// Instruction SLTIU behavior method.
 void ac_behavior(SLTIU) {
-  dbg_printf("SLTIU r%d, r%d, %d", rd, rs1, imm);
+  int imm;
+  imm = (imm4 << 11) | (imm3 << 5) | (imm2 << 1) | imm1;
+  dbg_printf("SLTIU r%d, r%d, %d\n", rd, rs1, imm);
   int sign_ext;
-  if ((imm >> 11) == 1)
+  if (imm4 == 1)
     sign_ext = imm | 0xFFFFF000;
   else
     sign_ext = imm;
   if ((ac_Uword)RB[rs1] < (ac_Uword)sign_ext)
     RB[rd] = 1;
   else
     RB[rd] = 0;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("Result = %#x\n\n", RB[rd]);
 }
 
+
+// Instruction XORI behavior method.
 void ac_behavior(XORI) {
-  dbg_printf("XORI r%d, r%d, %d", rd, rs1, imm);
+  int imm;
+  imm = (imm4 << 11) | (imm3 << 5) | (imm2 << 1) | imm1;
+  dbg_printf("XORI r%d, r%d, %d\n", rd, rs1, imm);
   int sign_ext;
-  if ((imm >> 11) == 1)
+  if (imm4 == 1)
     sign_ext = imm | 0xFFFFF000;
   else
     sign_ext = imm;
   RB[rd] = RB[rs1] ^ sign_ext;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  dbg_printf("Result = %#x\n\n", RB[rd]);
 }
 
+
+// Instruction ORI behavior method.
 void ac_behavior(ORI) {
-  dbg_printf("ORI r%d, r%d, %d", rd, rs1, imm);
+  int imm;
+  imm = (imm4 << 11) | (imm3 << 5) | (imm2 << 1) | imm1;
+  dbg_printf("ORI r%d, r%d, %d\n", rd, rs1, imm);
   int sign_ext;
   if ((imm >> 11) == 1)
     sign_ext = imm | 0xFFFFF000;
   else
     sign_ext = imm;
-  RB[rd] = RB[rs1] + sign_ext;
-  dbg_printf("Result = %#x\n", RB[rd]);
+  RB[rd] = RB[rs1] | sign_ext;
+  dbg_printf("Result = %#x\n\n", RB[rd]);
 }
 
+
+// Instruction ANDI behavior method.
 void ac_behavior(ANDI) {
-  dbg_printf("ANDI r%d, r%d, %d", rd, rs1, imm);
+  int imm;
+  imm = (imm4 << 11) | (imm3 << 5) | (imm2 << 1) | imm1;
+  dbg_printf("ANDI r%d, r%d, %d\n", rd, rs1, imm);
   int sign_ext;
   if ((imm >> 11) == 1)
     sign_ext = imm | 0xFFFFF000;
   else
     sign_ext = imm;
```

Comment:
Since this sign-extension code is so common throughout your implementation:

```
    int sign_ext;
    if ((imm >> 11) == 1)
      sign_ext = imm | 0xFFFFF000;
    else
      sign_ext = imm;
```

you could factor this out into a helper function, so it's easier to change it if we need to.